File: go/tree/binary_tree_preorder_traversal.go
package main

import (
	"container/list"
)

//二叉树的前序遍历

//递归

func preorderTraversal(root *TreeNode) []int {
	var res []int
	preHelper(root, &res)
	return res
}

func preHelper(node *TreeNode, res *[]int) {
	if node == nil {
		return
	}

	*res = append(*res, node.Val)
	preHelper(node.Left, res)
	preHelper(node.Right, res)
}

//迭代
func preorderTraversal2(root *TreeNode) []int {
	var res []int
	if root == nil {
		return res
	}

	stack := list.New()
	stack.PushFront(root)
	for stack.Len() > 0 {
		e := stack.Front()
		if e != nil {
			node := e.Value.(*TreeNode)
			res = append(res, node.Val)
			stack.Remove(e)

			if node.Right != nil {
				stack.PushFront(node.Right)
			}
			if node.Left != nil {
				stack.PushFront(node.Left)
			}
		}
	}
	return res
}

File: python/tree/leetocde/binary_tree_preorder_traversal.py
from typing import List


# 二叉树的前序遍历

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    # 递归
    def preorderTraversal_1(self, root: TreeNode) -> List[int]:
        res = []
        self.dfs(root, res)
        return res
       
    def dfs(self, node, res):
        if not node:
            return
        res.append(node.val)
        self.dfs(node.left, res)
        self.dfs(node.right, res)

    # 迭代
    def preorderTraversal_2(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        stack, res = [root], []
        while len(stack) > 0:
            node = stack.pop()
            res.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return res
